File: tools/parity/dispatch/agent-scratch-worktrees.ts
// TS dispatch for the agent-scratch-worktrees parity module. The shared TS impl
// was DELETED (`src/shared/agent-scratch-worktrees.ts` keeps only the two marker
// tables and the matcher type), so this drives the two places its bodies went.
//
// `isAgentScratchRepoRootPath` is the module's one seam SHIM
// (`src/shared/agent-scratch-repo-roots.ts`), driven here exactly as the app
// runs it: `config/vitest.parity.config.ts` installs no setup file, so the seam
// is unbound and the shim answers from its `parity` fallback — which is the
// deleted body, and the code main runs before its napi binding is installed.
//
// The two worktree matchers are deliberately NOT shims. Their only consumer is
// `worktree-ownership-rules.ts`, the non-dispatching pre-ready fallback of the
// already-cut-over `worktree-ownership-policy.ts`, so the bodies live there and
// this adapter calls them there. Routing them through a shim would compare Rust
// against Rust.
//
// `createAgentScratchWorktreePathMatcher` returned a closure, which no vector can
// carry back. Its arm therefore takes `{ checkoutPaths, worktreePath }` and
// answers the closure's verdict for that path — the per-row call
// `classifyWorktreeOwnership` makes, not a synthetic stand-in.

import { isAgentScratchRepoRootPath } from '../../../src/shared/agent-scratch-repo-roots'
import {
  legacyAgentScratchWorktreePathMatcher,
  legacyIsAgentScratchWorktreePath
} from '../../../src/shared/worktree-ownership-rules'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'isAgentScratchWorktreePath': {
      const { repoPath, worktreePath } = input as { repoPath: string; worktreePath: string }
      return legacyIsAgentScratchWorktreePath(repoPath, worktreePath)
    }
    case 'isAgentScratchRepoRootPath':
      return isAgentScratchRepoRootPath((input as { repoPath: string }).repoPath)
    case 'createAgentScratchWorktreePathMatcher': {
      const { checkoutPaths, worktreePath } = input as {
        checkoutPaths: string[]
        worktreePath: string
      }
      return legacyAgentScratchWorktreePathMatcher(checkoutPaths ?? [])(worktreePath)
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-dispatch/src/modules/agent_scratch_worktrees.rs
//! Parity dispatch for `orca_core::agent_scratch_worktrees` vs
//! `src/shared/agent-scratch-worktrees.ts`.
//!
//! `createAgentScratchWorktreePathMatcher` returns a closure, which cannot cross
//! a JSON boundary, so its arm takes `{ checkoutPaths, worktreePath }` and
//! answers the closure's result for that one path — the same shape both adapters
//! drive, and the same call `classifyWorktreeOwnership` makes per row.

use orca_core::agent_scratch_worktrees::{
    is_agent_scratch_repo_root_path, is_agent_scratch_worktree_path,
    AgentScratchWorktreePathMatcher,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "isAgentScratchWorktreePath" => Value::Bool(is_agent_scratch_worktree_path(
            &str_field(input, "repoPath"),
            &str_field(input, "worktreePath"),
        )),
        "isAgentScratchRepoRootPath" => {
            Value::Bool(is_agent_scratch_repo_root_path(&str_field(input, "repoPath")))
        }
        "createAgentScratchWorktreePathMatcher" => {
            let checkouts: Vec<String> = input
                .get("checkoutPaths")
                .and_then(Value::as_array)
                .map(|items| {
                    items.iter().map(|item| item.as_str().unwrap_or_default().to_string()).collect()
                })
                .unwrap_or_default();
            Value::Bool(
                AgentScratchWorktreePathMatcher::new(&checkouts)
                    .matches(&str_field(input, "worktreePath")),
            )
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}

fn str_field(value: &Value, key: &str) -> String {
    value.get(key).and_then(Value::as_str).unwrap_or("").to_string()
}
